internal/internal_worker_base.go

```diff
@@ -332,61 +347,100 @@ func (bw *baseWorker) runPoller() {
 	defer bw.stopWG.Done()
 	bw.metricsHandler.Counter(metrics.PollerStartCounter).Inc(1)
 
+	ctx, cancelfn := context.WithCancel(context.Background())
+	reserveChan := make(chan *SlotPermit)
+
 	for {
+		bw.stopWG.Add(1)
+		go func() {
+			defer bw.stopWG.Done()
+			s, err := bw.slotSupplier.ReserveSlot(ctx, &bw.options.slotReservationData)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					bw.logger.Error(fmt.Sprintf("Error while trying to reserve slot: %v", err))
+				} else {
+					close(reserveChan)
+				}
+				return
+			}
+			select {
+			case reserveChan <- s:
+			case <-ctx.Done():
+				bw.releaseSlot(s, "worker shutting down")
+			}
+		}()
+
 		select {
 		case <-bw.stopCh:
+			cancelfn()
 			return
-		case <-bw.pollerRequestCh:
+		case permit := <-reserveChan:
+			if permit == nil { // There was an error reserving a slot
+				// Avoid spamming reserve hard in the event it's constantly failing
+				time.Sleep(time.Second)
+				continue
+			}
 			if bw.sessionTokenBucket != nil {
 				bw.sessionTokenBucket.waitForAvailableToken()
 			}
-			bw.pollTask()
+			bw.pollTask(permit)
 		}
 	}
 }
 
-func (bw *baseWorker) tryReserveSlot() bool {
+func (bw *baseWorker) tryReserveSlot() *SlotPermit {
 	if bw.isStop() {
-		return false
-	}
-	// Reserve a executor slot via a non-blocking attempt to take a poller
-	// request entry which essentially reserves a slot
-	select {
-	case <-bw.pollerRequestCh:
-		return true
-	default:
-		return false
+		return nil
 	}
+	return bw.slotSupplier.TryReserveSlot(&bw.options.slotReservationData)
 }
 
-func (bw *baseWorker) releaseSlot() {
-	// Like other sends to this channel, we assume there is room because we
-	// reserved it, so we make a blocking send.
-	bw.pollerRequestCh <- struct{}{}
+func (bw *baseWorker) releaseSlot(permit *SlotPermit, reason string) {
+	bw.slotSupplier.ReleaseSlot(permit, reason)
 }
 
 func (bw *baseWorker) pushEagerTask(task eagerTask) {
-	// Should always be non blocking if a slot was reserved.
+	// Should always be non-blocking. Slots are reserved before requesting eager tasks.
 	bw.eagerTaskQueueCh <- task
 }
 
-func (bw *baseWorker) processTaskAsync(task interface{}, callback func()) {
+func (bw *baseWorker) processTaskAsync(eagerOrPolled eagerOrPolledTask) {
 	bw.stopWG.Add(1)
 	go func() {
-		if callback != nil {
-			defer callback()
+		defer bw.stopWG.Done()
+
+		task := eagerOrPolled.getTask()
+		permit := eagerOrPolled.getPermit()
+
+		if !task.isEmpty() {
+			bw.slotSupplier.MarkSlotUsed(permit)
+		}
+
+		defer func() {
+			bw.releaseSlot(permit, "task processed")
+
+			if p := recover(); p != nil {
```

Comment:
Hrmm. Did you see a situation where a panic happened, or is this a result of some other refactoring?

Reply:
It's from inlining the `processTask` function that was only ever called from here